### PR TITLE
Dev/vm runner initial

### PR DESCRIPTION
**Pull request by JuanBesa**
This is the Basic Execution Engine code and tests that shows how to run etch code. It is the core execution engine which the other interfaces use.

### CODE REVIEW

File: libs/dmlf/include/dmlf/local_vm_launcher.hpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+class LocalVmLauncher : public VmLauncherInterface
+{
+public:
+  LocalVmLauncher()           = default;
+  ~LocalVmLauncher() override = default;
+
+  LocalVmLauncher(const LocalVmLauncher &other) = delete;
+  LocalVmLauncher &operator=(const LocalVmLauncher &other)  = delete;
+  bool             operator==(const LocalVmLauncher &other) = delete;
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/local_vm_launcher.hpp]: == operator and comparison operator are implicitly not implemented so you don’t need to mark them as explicitly as such

File: libs/dmlf/include/dmlf/local_vm_launcher.hpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  bool CreateState(std::string name) override;
+  bool HasState(std::string const &name) const override;
+  bool CopyState(std::string const &srcName, std::string newName) override;
+
+  bool Execute(std::string const &programName, std::string const &vmName,
+               std::string const &stateName, std::string const &entrypoint,
+               Params const &params) override;
+  void AttachExecuteErrorHandler(ExecuteErrorHandler newHandler) override;
+
+  std::unordered_map<std::string, std::shared_ptr<VM>> vms_;
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/local_vm_launcher.hpp]: Worth adding some type aliases for these containers? Also not sure if we want this data structure public?

File: libs/dmlf/include/dmlf/vm_launcher_interface.hpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+class VmLauncherInterface
+{
+public:
+  VmLauncherInterface()          = default;
+  virtual ~VmLauncherInterface() = default;
+
+  VmLauncherInterface(const VmLauncherInterface &other) = delete;
+  VmLauncherInterface &operator=(const VmLauncherInterface &other) = delete;
+
+  bool operator==(const VmLauncherInterface &other) = delete;
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/vm_launcher_interface.hpp]: See comment above

File: libs/dmlf/include/dmlf/vm_launcher_interface.hpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+namespace fetch {
+namespace dmlf {
+
+class VmLauncherInterface
+{
+public:
+  VmLauncherInterface()          = default;
+  virtual ~VmLauncherInterface() = default;
+
+  VmLauncherInterface(const VmLauncherInterface &other) = delete;
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/vm_launcher_interface.hpp]: Normally for a purely abstract class you would defer the non copyable-ness to the concrete implementation

File: libs/dmlf/include/dmlf/vm_persistent.hpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  VmPersistent(const VmPersistent &other) = delete;
+  VmPersistent &operator=(const VmPersistent &other) = delete;
+
+  Status Read(std::string const &key, void *data, uint64_t &size) override;
+  Status Write(std::string const &key, void const *data, uint64_t size) override;
+  Status Exists(std::string const &key) override;
+
+  VmPersistent DeepCopy() const;
+
+  bool operator==(const VmPersistent &other) = delete;
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/vm_persistent.hpp]: See above

File: libs/dmlf/src/local_vm_launcher.cpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  {
+    executeErrorhandler_(programName, vmName, stateName, runTimeError);
+    return false;
+  }
+
+  return thereWasAnError;
+}
+void LocalVmLauncher::AttachExecuteErrorHandler(ExecuteErrorHandler newHandler)
+{
+  executeErrorhandler_ = newHandler;
```
Comment (ejfitzgerald) [inline libs/dmlf/src/local_vm_launcher.cpp]: I can't see from the review but presumably:

```suggestion
  executeErrorhandler_ = std::move(newHandler);
```

**Review by ejfitzgerald — COMMENTED** [review]
Looks good to me just some minor (read pedantic 😁 ) points

File: libs/dmlf/include/dmlf/vm_launcher_interface.hpp [hunk 7]
```diff
@@ -55,9 +62,9 @@ class VmLauncherInterface
   virtual bool HasVM(std::string const &name) const                      = 0;
   virtual bool SetVmStdout(std::string const &vmName, VmOutputHandler &) = 0;
 
-  virtual bool CreateState(std::string name)                              = 0;
-  virtual bool HasState(std::string const &name) const                    = 0;
-  virtual bool CopyState(std::string const &srcName, std::string newName) = 0;
+  virtual bool CreateState2(std::string name)                              = 0;
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/vm_launcher_interface.hpp]: CreateState2?

File: libs/dmlf/tests/execution/basic_engine_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,860 @@
...
+  EXPECT_EQ(result.output().Get<int>(), 2);
+
+  result = engine.Run("tick", "state2", "main");
+  EXPECT_TRUE(result.succeeded());
+  EXPECT_EQ(result.output().Get<int>(), 1);
+}
+
+// Repeated things
+
+// Breaks VM
```
Comment (ejfitzgerald) [inline libs/dmlf/tests/execution/basic_engine_tests.cpp:252]: 😢

File: libs/dmlf/include/dmlf/local_vm_launcher.hpp (hunk 2 above)
Comment (JuanBesa) [reply libs/dmlf/include/dmlf/local_vm_launcher.hpp]: Moved to private. I think this is the only place it is declared so am not sure a typedef is necessary.

File: libs/dmlf/include/dmlf/vm_launcher_interface.hpp (hunk 4 above)
Comment (JuanBesa) [reply libs/dmlf/include/dmlf/vm_launcher_interface.hpp]: This class was deleted. I resolved it in the ExecutionEngineInterface class

File: libs/dmlf/include/dmlf/vm_launcher_interface.hpp (hunk 3 above)
Comment (JuanBesa) [reply libs/dmlf/include/dmlf/vm_launcher_interface.hpp]: Done!

File: libs/dmlf/include/dmlf/vm_launcher_interface.hpp (hunk 7 above)
Comment (JuanBesa) [reply libs/dmlf/include/dmlf/vm_launcher_interface.hpp]: This class should have been deleted, I removed it (the CreateState2 was an artefact left over from the transition)

File: libs/dmlf/tests/execution/basic_engine_tests.cpp (hunk 8 above)
Comment (JuanBesa) [reply libs/dmlf/tests/execution/basic_engine_tests.cpp:252]: Yeah Robert knows and told he would look into a fix

**Review by engineerkatie — APPROVED** [review]

**Pull request merged**